### ISSUE TITLE
Pyright incorrectly narrows the type on `issubclass`, which can result in false positive `UnnecessaryIsInstance`

### ISSUE THREAD

**Issue by kennipj** [microsoft/pyright#1851]
Pyright incorrectly narrows the type on `issubclass`, which also results in false positive `UnnecessaryIsInstance` error when checking for multi-class inheritance.

Code sample:
```py
from inspect import isclass
from typing import Any


class A:
    pass


class B:
    pass


def some_func(obj: Any) -> None:
    if isclass(obj) and issubclass(obj, A):
        reveal_type(obj)  # Type of "obj" is "Type[A]" (mypy: Revealed type is 'Any')
        if issubclass(obj, B):
            reveal_type(obj)  # Type of "obj" is "Never"
            # Unnecessary issubclass call; "Type[A]" is never a subclass of "B"
            pass
```
In this case `obj` can obviously also implement `B`, but pyright incorrectly narrows the type to be only `Type[A]`. When the type should probably be something like `Type[A & Unknown]` (although there is no python equivalent)

I'm aware that python does not currently support an intersection of types, but only `Union` (relevant discussion: https://github.com/python/typing/issues/213). However it would be useful if pyright could narrow this to a pyright specific `Type[A & B & Unknown]`.

Regardless, I would not expect pyright to report an `UnnecessaryIsInstance` error in this case.

Mypy does not seem to narrow the type at all (although I would highly prefer proper type narrowing in this case, rather than just not doing it all).

Tested on:
- pyright `1.1.138`
- mypy `0.812`

**Comment by erictraut** [comment]
I see your point that multiple nested `issubclass` checks (or `isinstance` checks) may not compose because of the possibility of multiple inheritance. 

Introducing an "Unknown" during type narrowing will result in false positive errors in strict mode, so I don't think that's a good idea. Introducing an intersection type into the Python type system just to address this one case isn't feasible. It would introduce a new typing concept to Python that doesn't currently exist, and it would require significant changes to the type checker and would generally impact overall type analyzer performance for little benefit.

Also, your proposal would require that pyright track the full history of how a type was narrowed, something it doesn't do currently. That would be required to distinguish between the case where nested `issubclass` checks are used, versus a single check. This would also be really complex and expensive.

I'll think about this more, but I don't see an immediate, obvious solution to this problem, at least none that don't involve massive costs in complexity and type analyzer performance.

This might be a limitation that we just need to live with. The `reportUnnecessaryIsInstance` check is an advanced rule that is off by default. If you don't like its behavior in this particular edge case, you have the option of not enabling it (or disabling it in specific cases).

There are probably other workarounds depending on the particulars of your use case, but I'm guessing that your situation is more complex than the simplified code sample indicates.

Like I said, I'll think about this more. Maybe I'll come up with some other solution after sleeping on it.

**Comment by kennipj** [comment]
I suppose this is really only an issue when trying to narrow type from `Any`. Since in this case `obj` could implement any number of classes. So maybe the way to solve this, is to just not do type narrowing at all in this case, since you can't guarantee the ordering of the implemented classes, so any field/method you access could be something unknown.

**Comment by erictraut** [comment]
I think this applies to cases other than `Any` because some class `C` could theoretically subclass from both `A` and `B`. So if the object is known to be of type `A`, it could be argued that `issubclass(object, B)` should not narrow to `Never`. However, this is an edge case, and there's utility in narrowing to `Never` in most cases. The entire `reportUnnecessaryIsInstance` check would be impossible if we made the assumption that any variable is possibly a subclass of any type.

Likewise, narrowing `Any` in an `isinstance` or `issubclass` type guard is _very_ useful most of the time. It allows one to get meaningful type checking and completion suggestions within the guarded block even if the type of the variable was not known (or not specified) outside of the block. If we were to change the behavior as you suggest and never narrow `Any` in this case, we'd be giving up these very useful behaviors. I don't think that's a good tradeoff.

**Comment by erictraut** [comment]
I've updated the `isinstance` and `issubclass` type narrowing logic to synthesize an "intersection" type in this situation. I've also updated the `reportUnnecessaryIsInstance` diagnostic check to report only the cases where the test is always true. The always-false case is no longer reported as an error.

In the code sample at the start of this issue, pyright now reports the following types:

```python
def some_func(obj: Any) -> None:
    if isclass(obj) and issubclass(obj, A):
        reveal_type(obj)  # Type of "obj" is "Type[A]"
        if issubclass(obj, B):
            reveal_type(obj)  # Type of "obj" is "Type[<subclass of A and B>]"
            pass
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.155, which I just published. It will also be included in the next release of pylance.
